Sources/SourceKitLSP/Swift/SyntaxHighlightingToken.swift

```diff
@@ -0,0 +1,413 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SourceKitD
+import LanguageServerProtocol
+import LSPLogging
+
+/// A ranged token in the document used for syntax highlighting.
+public struct SyntaxHighlightingToken: Hashable {
+  public var start: Position
+  public var length: Int
+  public var kind: Kind
+  public var modifiers: Modifiers
+
+  /// The end of a token. Note that this requires the token to be
+  /// on a single line, which is the case for all tokens emitted
+  /// by parseTokens, however.
+  public var sameLineEnd: Position {
+    Position(line: start.line, utf16index: start.utf16index + length)
+  }
+  public var sameLineRange: Range<Position> {
+    start..<sameLineEnd
+  }
+
+  public init(
+    name: String? = nil,
+    start: Position,
+    length: Int,
+    kind: Kind,
+    modifiers: Modifiers = []
+  ) {
+    self.start = start
+    self.length = length
+    self.kind = kind
+    self.modifiers = modifiers
+  }
+
+  /// Splits a potentially multi-line token to multiple single-line tokens.
+  public func splitToSingleLineTokens(in snapshot: DocumentSnapshot) -> [Self]? {
+    guard let startIndex = snapshot.index(of: start) else {
```

Comment:
I think it would be better to crash here instead of silently dropping the token. If the token is not in the snapshot, that’s a serious violation of our assumptions and I’d rather get crash reports about it than vague bug reports saying that tokens are not being highlighted.